src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -758,10 +759,17 @@ void executeSuppressionActions(IEnumerable<Diagnostic> reportedDiagnostics, bool
                 var suppressors = this.Analyzers.OfType<DiagnosticSuppressor>();
                 if (concurrent)
                 {
-                    Parallel.ForEach(suppressors, suppressor =>
+                    try
                     {
-                        AnalyzerExecutor.ExecuteSuppressionAction(suppressor, getSuppressableDiagnostics(suppressor));
-                    });
+                        Parallel.ForEach(suppressors, suppressor =>
+                        {
+                            AnalyzerExecutor.ExecuteSuppressionAction(suppressor, getSuppressableDiagnostics(suppressor));
+                        });
+                    }
+                    catch (AggregateException ex)
+                    {
+                        ExceptionDispatchInfo.Capture(ex.InnerException).Throw();
```

Comment:
>InnerException [](start = 57, length = 14)

What if there are exceptions from multiple threads?

Reply:
Should `OperationCanceledException` reach here in an `AggregateException`? If `Parallel.ForEach()` was passed a `ParallelOptions` with a `CancellationToken`, then we'd presumably have to handle non-nested `OperationCanceledException` here. But otherwise, should the `Action` handle the `OperationCanceledException` for each thread?

---
In reply to: [383568522](https://github.com/dotnet/roslyn/pull/41892#discussion_r383568522) [](ancestors = 383568522)

Reply:
Good points. @sharwell Do you know how `ExceptionDispatchInfo.Capture` handles these cases? I think it might be best to switch this code to just building up a tasks array and doing a WaitAll instead of using a `Parallel.ForEach`, just like other places in the driver. Another option is to remove the concurrent execution completely as we don't expect that many suppressors, and they would generally be much cheaper compared to analyzers.

Reply:
@cston @sharwell I have switched to a different approach.